### PR TITLE
Update AccountTracker methods to be #private

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**

* Reduces the API surface of AccountTracker by enforcing private methods via `#` prefix
* Updates specs
  * Notably, merges old `_updateAccount` and `_updateAccountViaBalanceChecker` into `updateAccounts`
* Fixes issue with state object being updated directly by using cloneDeep

## **Related issues**

Fixes:

## **Manual testing steps**

Nothing should have changed for the user

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by jiexi** [comment]
Cleaned up the AccountTracker specs here https://github.com/MetaMask/metamask-extension/pull/22233

File: app/scripts/lib/account-tracker.test.js [hunk 1]
```diff
@@ -146,12 +148,12 @@ describe('Account Tracker', () => {
       expect(blockTrackerStub.removeListener).toHaveBeenNthCalledWith(
         1,
         'latest',
-        accountTracker._updateForBlock,
+        expect.any(Function),
```
Comment (adonesky1) [inline app/scripts/lib/account-tracker.test.js:151]: Is it possible to make keep this more specific?

File: app/scripts/lib/account-tracker.js [hunk 2]
```diff
@@ -382,9 +394,9 @@ export default class AccountTracker {
    * @param {number} blockNumber - the block number to update to.
    * @fires 'block' The updated state, if all account updates are successful
    */
-  async _updateForBlock(blockNumber) {
-    await this._updateForBlockByNetworkClientId(null, blockNumber);
-  }
+  #updateForBlock = async (blockNumber) => {
```
Comment (adonesky1) [inline app/scripts/lib/account-tracker.js:397]: [nit] wondering why this was made into an arrow function. Now its the only one on the controller

File: app/scripts/lib/account-tracker.js [hunk 3]
```diff
@@ -586,25 +593,36 @@ export default class AccountTracker {
...
+          ...accountsByChainId,
+          [chainId]: newAccounts,
+        },
+      });
     } catch (error) {
       log.warn(
         `MetaMask - Account Tracker single call balance fetch failed`,
         error,
       );
       Promise.all(
```
Comment (adonesky1) [inline app/scripts/lib/account-tracker.js:623]: Another chance here to revisit whether this should be Promise.all vs Promise.allSettled. I think the latter. @shanejonas agrees: https://github.com/MetaMask/metamask-extension/pull/21868#discussion_r1412386135

Comment (adonesky1) [reply app/scripts/lib/account-tracker.js:623]: cc @mcmire @Gudahtt any thoughts?

File: app/scripts/lib/account-tracker.test.js [hunk 4]
```diff
@@ -39,6 +37,10 @@ const mockAccounts = {
   },
 };
 
+function flushPromises() {
+  return new Promise(jest.requireActual('timers').setImmediate);
+}
+
```
Comment (adonesky1) [inline app/scripts/lib/account-tracker.test.js:43]: This is now a duplicate (also used in detect-tokens.js) we should pull it out into a helper file with some comments about what its doing.

File: app/scripts/lib/account-tracker.test.js [hunk 5]
```diff
@@ -451,114 +453,16 @@ describe('Account Tracker', () => {
...
-        },
-        currentBlockGasLimit: '',
-        currentBlockGasLimitByChainId: {},
-      });
-    });
-  });
-
-  describe('_updateForBlockByNetworkClientId', () => {
+  describe('blockTracker "latest" events', () => {
     it('updates currentBlockGasLimit, currentBlockGasLimitByChainId, and accounts when no networkClientId is passed', async () => {
```
Comment (adonesky1) [inline app/scripts/lib/account-tracker.test.js:457]:
```suggestion
    it('updates currentBlockGasLimit, currentBlockGasLimitByChainId, and accounts when polling is initiated via `start`', async () => {
```

File: app/scripts/lib/account-tracker.test.js [hunk 6]
```diff
@@ -572,17 +476,20 @@ describe('Account Tracker', () => {
           [currentChainId]: GAS_LIMIT,
         },
       });
+
+      accountTracker.stop();
     });
 
     it('updates only the currentBlockGasLimitByChainId and accounts when a networkClientId is passed', async () => {
```
Comment (adonesky1) [inline app/scripts/lib/account-tracker.test.js:483]:
```suggestion
    it('updates only the currentBlockGasLimitByChainId and accounts when polling is initiated via `startPollingByNetworkClientId`', async () => {
```

**Review by adonesky1 — COMMENTED** [review]
Looking good. Just a few small comments

File: app/scripts/lib/account-tracker.test.js (hunk 1 above)
Comment (jiexi) [reply app/scripts/lib/account-tracker.test.js:151]: I'm not sure how as the `_updateForBlock` method is now private and inaccessible from the test

File: app/scripts/lib/account-tracker.test.js (hunk 4 above)
Comment (jiexi) [reply app/scripts/lib/account-tracker.test.js]: done https://github.com/MetaMask/metamask-extension/pull/22217/commits/873f6671e2d4be7af44ecb35fd7e13fbf630e1ad

Comment (jiexi) [reply app/scripts/lib/account-tracker.test.js]: linter doesn't like `jest` being used in a non *.test file. We can pass it in as an arg to flushPromises? seems weird though

The docs actually mention this function in it's entirety
https://github.com/MetaMask/metamask-extension/blob/develop/docs/testing.md?plain=1#L160

Not sure what the best approach is here

Comment (jiexi) [reply app/scripts/lib/account-tracker.test.js]: fixed lint issue here https://github.com/MetaMask/metamask-extension/pull/22217/commits/e24e829eef6c29832043fd592144d2852588d277

**Review by adonesky1 — DISMISSED** [review]
LGTM!

File: app/scripts/lib/account-tracker.js (hunk 3 above)
Comment (mcmire) [reply app/scripts/lib/account-tracker.js:623]: Yes, that's a good call.

Comment (jiexi) [reply app/scripts/lib/account-tracker.js]: done https://github.com/MetaMask/metamask-extension/pull/22217/commits/e4a781096b6d36d67b8f15a9567ec190e3ba3dd5

File: app/scripts/lib/account-tracker.js [hunk 7]
```diff
@@ -33,10 +34,10 @@ import { previousValueComparator } from './util';
  * @property {object} store.accountsByChainId The accounts currently stored in this AccountTracker keyed by chain id
  * @property {string} store.currentBlockGasLimit A hex string indicating the gas limit of the current block
  * @property {string} store.currentBlockGasLimitByChainId A hex string indicating the gas limit of the current block keyed by chain id
- * @property {object} _provider A provider needed to create the EthQuery instance used within this AccountTracker.
- * @property {BlockTracker} _blockTracker A BlockTracker instance. Needed to ensure that accounts and their info updates
+ * @property {object} #provider A provider needed to create the EthQuery instance used within this AccountTracker.
+ * @property {BlockTracker} #blockTracker A BlockTracker instance. Needed to ensure that accounts and their info updates
  * when a new block is created.
```
[the remark this answers is not in the document]
Comment (jiexi) [reply app/scripts/lib/account-tracker.js]: Done https://github.com/MetaMask/metamask-extension/pull/22217/commits/6619babaec1aa9273b52feff23c7a6ac494a4eb2

File: app/scripts/lib/account-tracker.js [hunk 8]
```diff
@@ -33,10 +34,10 @@ import { previousValueComparator } from './util';
...
- * @property {object} _currentBlockNumberByChainId Reference to a property on the _blockTracker: the number (i.e. an id) of the the current block keyed by chain id
+ * @property {object} _currentBlockNumberByChainId Reference to a property on the #blockTracker: the number (i.e. an id) of the the current block keyed by chain id
```
[the remark this answers is not in the document]
Comment (jiexi) [reply app/scripts/lib/account-tracker.js]: The actual method isn't hashtag private. I've updated that below and removed this jsdoc property

done https://github.com/MetaMask/metamask-extension/pull/22217/commits/c17b29a753ef771d1c795d8fba8d66600be929bc

**Review by adonesky1 — APPROVED** [review]

File: app/scripts/lib/account-tracker.js (hunk 2 above)
Comment (shanejonas) [reply app/scripts/lib/account-tracker.js:395]: `=` would make it a static method on the class

Comment (jiexi) [reply app/scripts/lib/account-tracker.js:395]: i don't believe that is true. I think the `static` keyword defines that

File: app/scripts/lib/account-tracker.js [hunk 9]
```diff
@@ -586,25 +591,36 @@ export default class AccountTracker {
...
+          [chainId]: newAccounts,
+        },
+      });
     } catch (error) {
       log.warn(
         `MetaMask - Account Tracker single call balance fetch failed`,
         error,
       );
-      Promise.all(
+      Promise.allSettled(
```
Comment (mcmire) [inline app/scripts/lib/account-tracker.js:621]: What are your thoughts on moving this to a separate PR? This seems unrelated to using `#` syntax for private methods, and might get lost here.

File: app/scripts/lib/account-tracker.js [hunk 10]
```diff
@@ -347,7 +354,10 @@ export default class AccountTracker {
    * @param {Array} addresses - An array of hex addresses to stop tracking.
    */
   removeAccounts(addresses) {
-    const { accounts, accountsByChainId } = this.store.getState();
+    const { accounts: _accounts, accountsByChainId: _accountsByChainId } =
```
Comment (mcmire) [inline app/scripts/lib/account-tracker.js:357]: What are your thoughts on moving this to a separate PR? This seems unrelated to switching to `#` syntax for private methods.

File: app/scripts/lib/account-tracker.test.js [hunk 11]
```diff
@@ -451,114 +450,16 @@ describe('Account Tracker', () => {
     });
   });
 
-  describe('_updateAccount', () => {
```
Comment (mcmire) [inline app/scripts/lib/account-tracker.test.js:454]: Just want to confirm: Have we checked tests for methods that call `_updateAccount` to see whether they include these?

File: app/scripts/lib/account-tracker.test.js [hunk 12]
```diff
@@ -908,75 +829,4 @@ describe('Account Tracker', () => {
       });
     });
   });
-
-  describe('_updateAccountsViaBalanceChecker', () => {
```
Comment (mcmire) [inline app/scripts/lib/account-tracker.test.js:912]: Same question here — do the tests elsewhere cover the behavior that was previously tested here?

**Review by mcmire — COMMENTED** [review]
I did a quick pass over this PR, but will take some time to look at it closely tomorrow if it's still open.

File: app/scripts/lib/account-tracker.test.js (hunk 11 above)
Comment (jiexi) [reply app/scripts/lib/account-tracker.test.js:454]: yes, this got folded into the scenarios in updateAccount

File: app/scripts/lib/account-tracker.test.js (hunk 12 above)
Comment (jiexi) [reply app/scripts/lib/account-tracker.test.js:912]: yup! covered in updateAccount now. Different entrypoint, but equivalent testing

**Review by shanejonas — APPROVED** [review]

File: app/scripts/lib/account-tracker.js (hunk 9 above)
Comment (jiexi) [reply app/scripts/lib/account-tracker.js:621]: Let's get this merged first then https://github.com/MetaMask/metamask-extension/pull/22306

File: app/scripts/lib/account-tracker.js (hunk 10 above)
Comment (jiexi) [reply app/scripts/lib/account-tracker.js:357]: Done. Let's get this merged first then https://github.com/MetaMask/metamask-extension/pull/22306
